pkg/controllers/user/multiclusterapp/controller.go

```diff
@@ -0,0 +1,114 @@
+package multiclusterapp
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	pv3 "github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/rancher/types/config"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	globalScopeAnswersKey = "global"
+)
+
+type MultiClusterAppController struct {
+	MultiClusterAppClient v3.MultiClusterAppInterface
+	MultiClusterAppLister v3.MultiClusterAppLister
+	AppClient             pv3.AppInterface
+	//AppRevisionGetter     v3.AppRevisionsGetter
+	//AppGetter             v3.AppsGetter
+	NsClient corev1.NamespaceInterface
+	NsLister corev1.NamespaceLister
+}
+
+func Register(ctx context.Context, cluster *config.UserContext) {
+	m := &MultiClusterAppController{
+		//AppGetter:             user.Management.Project,
+		//AppRevisionGetter:     user.Management.Project,
+		AppClient:             cluster.Management.Project.Apps(""),
+		NsClient:              cluster.Management.Core.Namespaces(""),
+		NsLister:              cluster.Management.Core.Namespaces("").Controller().Lister(),
+		MultiClusterAppClient: cluster.Management.Management.MultiClusterApps(""),
+		MultiClusterAppLister: cluster.Management.Management.MultiClusterApps("").Controller().Lister(),
+	}
+	m.MultiClusterAppClient.AddHandler(ctx, "multi-cluster-app-controller", m.sync)
+}
+
+func (m *MultiClusterAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
+	if mcapp == nil || mcapp.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	var answerMap map[string]map[string]string
+	var answerFound bool
+
+	// create app object spec
+	// create the externalId field
+	externalId := fmt.Sprintf("catalog://?catalog=%s&template=%s&version=%s", mcapp.Spec.CatalogName, mcapp.Spec.TemplateName, mcapp.Spec.TemplateVersionName)
+
+	// create a map, where key is the projectID or clusterID, or "global" if neither is provided, and value is the actual answer values
+	answerMap = make(map[string]map[string]string)
+	for _, a := range mcapp.Spec.Answers {
+		if a.ProjectName != "" {
+			answerMap[a.ProjectName] = make(map[string]string)
+			answerMap[a.ProjectName] = a.Values
+		} else if a.ClusterName != "" {
+			answerMap[a.ClusterName] = make(map[string]string)
+			answerMap[a.ClusterName] = a.Values
+		} else {
+			answerMap[globalScopeAnswersKey] = make(map[string]string)
+			answerMap[globalScopeAnswersKey] = a.Values
+		}
+	}
+
+	// for all targets, create the App{} instance, so that helm controller App lifecycle can pick it up
+	for _, t := range mcapp.Spec.Targets {
+		app := pv3.App{
+			Spec: pv3.AppSpec{
+				ProjectName:         t.Spec.ProjectName,
```

Comment:
we need to add an owner reference to the multi cluster app. Otherwise app won't get cleaned up when multi cluster app gets removed

Reply:
Discussed offline that since ownerReference won't work cross namespace, we'll add a method in sync for now to delete the apps if the multiclusterapp object passed is nil or has DeletionTimestamp set